### ISSUE TITLE
Object rest inside a loop declaration

### ISSUE THREAD

**Issue by peey** [babel/babylon#550]
If an assignment expression has a object pattern which has a `restElement`, then babylon thinks it's a syntax error when it shouldn't be
### Input Code

```js
var o = {"i": 3, "rest": "foo"};
for (var {i, ...rest} = o; i <= 10; i++) {
 console.log(rest);
}
```

[repl link](https://babeljs.io/repl/#?babili=false&evaluate=true&lineWrap=false&presets=es2015&targets=Node-7.9&browsers=&builtIns=false&debug=false&code=var%20o%20%3D%20%7B%22i%22%3A%203%2C%20%22rest%22%3A%20%22foo%22%7D%3B%0Afor%20(var%20%7Bi%2C%20...rest%7D%20%3D%20o%3B%20i%20%3C%3D%2010%3B%20i%2B%2B)%20%7B%0A%20console.log(rest)%3B%0A%7D&experimental=true&loose=true&spec=false&playground=false&stage=0)
### Configuration

Babel REPL configuration and environment

**Comment by hzoo** [comment]
Hey @peey! We really appreciate you taking the time to report an issue. The collaborators
on this project attempt to help as many people as possible, but we're a limited number of volunteers,
so it's possible this won't be addressed swiftly.

If you need any help, or just have general Babel or JavaScript questions, we have a vibrant Slack
community that typically always has someone willing to help. You can sign-up [here](https://slack.babeljs.io/)
for an invite.

**Comment by peey** [comment]
I forgot to mention that this problem is also present in while and do-while loops

```js
let o = {x: "ex", y: "wai", z: "zed"};
while({x, y, ...z} = o) {
  // code
}

do {
  //something
}
while({x, y, ...z} = o);
```
But the same code without the `restElement` `...z` works fine. I think to solve this we should allow a `restElement` wherever an `objectPattern` exists?

**Comment by jridgewell** [comment]
This was fixed in at some point in 7.0 branch.
